### PR TITLE
Add user avatar

### PR DESCRIPTION
**Pull request by spiderjako**

### CODE REVIEW

File: lib/sanbase/auth/user.ex [hunk 1]
```diff
@@ -167,6 +170,13 @@ defmodule Sanbase.Auth.User do
     end
   end
 
+  defp validate_url_change(_, url) do
+    case Sanbase.Validation.valid_url?(url) do
+      :ok -> []
+      {:error, reason} -> [icon_url: reason]
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: copy-paste

File: lib/sanbase/auth/user.ex [hunk 2]
```diff
@@ -402,6 +412,12 @@ defmodule Sanbase.Auth.User do
   def insights_fallback_email, do: @insights_fallback_email
   def sanbase_bot_email, do: @sanbase_bot_email
 
+  def update_avatar_url(%User{avatar_url: avatar_url} = user) do
+    user
+    |> change(avatar_url: avatar_url)
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: This won't go through the changeset function which does the validation, right?

File: test/sanbase/auth/user_test.exs [hunk 3]
```diff
@@ -400,4 +400,15 @@ defmodule Sanbase.Auth.UserTest do
...
+  test "user with avatar url is okay" do
+    avatar_url =
+      "http://stage-sanbase-images.s3.amazonaws.com/uploads/_empowr-coinHY5QG72SCGKYWMN4AEJQ2BRDLXNWXECT.png"
+
+    user =
+      insert(:user)
+      |> Sanbase.Auth.User.update_avatar_url(avatar_url)
+
+    assert user.avatar_url == avatar_url
+  end
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: Add tests with not valid URLs, too

File: test/sanbase/auth/user_test.exs [hunk 4]
```diff
@@ -400,4 +400,25 @@ defmodule Sanbase.Auth.UserTest do
...
+
+  test "user with invalid avatar url returns proper error" do
+    avatar_url = "something invalid"
+
+    {:error, msg} =
+      insert(:user, avatar_url: avatar_url)
+      |> Sanbase.Auth.User.update_avatar_url()
+
+    assert msg == "`something invalid` is missing scheme"
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: this message is not very descriptive

File: lib/sanbase/auth/user.ex [hunk 5]
```diff
@@ -167,6 +170,13 @@ defmodule Sanbase.Auth.User do
     end
   end
 
+  defp validate_url_change(_, url) do
+    case Sanbase.Validation.valid_url?(url) do
+      :ok -> []
+      {:error, msg} -> [avatar_url: msg]
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: ignore the message and just say `"#{isnpect(url)} is not a vallid url"`. As seen from the test the error message here is `is missing scheme` which is not clear

File: lib/sanbase/auth/user.ex [hunk 6]
```diff
@@ -410,6 +420,18 @@ defmodule Sanbase.Auth.User do
     String.replace(@sanbase_bot_email, "@", "#{idx}@")
   end
 
+  def update_avatar_url(%User{} = user, avatar_url) do
+    case validate_url_change(user, avatar_url) do
+      [] ->
+        user
+        |> changeset(%{avatar_url: avatar_url})
+        |> Repo.update()
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: part of the changeset job is to validate, no point of doing that yourself

File: lib/sanbase_web/graphql/resolvers/user/user_resolver.ex [hunk 7]
```diff
@@ -158,6 +158,18 @@ defmodule SanbaseWeb.Graphql.Resolvers.UserResolver do
     end
   end
 
+  def change_avatar(_root, %{avatar_url: avatar_url}, %{
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/user/user_resolver.ex]: You also need to define the mutation in `user_queries.ex`

Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/resolvers/user/user_resolver.ex]: When you add a migration, add tests in the appropriate test/sanbase_web/... location

File: lib/sanbase_web/graphql/resolvers/user/user_resolver.ex [hunk 8]
```diff
@@ -158,6 +158,18 @@ defmodule SanbaseWeb.Graphql.Resolvers.UserResolver do
...
+        context: %{auth: %{auth_method: :user_token, current_user: user}}
+      }) do
+    User.update_avatar_url(user, avatar_url)
+    |> case do
+      {:ok, user} ->
+        {:ok, user}
+      {:error, _msg} ->
+        {:error, "Cannot update current user's avatar to #{inspect(avatar_url)}"}
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/user/user_resolver.ex]: I think the error message is good enough and you can just leave `User.update_avatar_url(user, avatar_url)` here

File: test/sanbase/auth/user_test.exs [hunk 9]
```diff
@@ -400,4 +400,25 @@ defmodule Sanbase.Auth.UserTest do
       assert length(User.users_with_monitored_watchlist_and_email()) == 0
     end
   end
+
+  test "user with avatar url is okay" do
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: Not sure what this description means

`can set valid avatar url` would be better

Comment (spiderjako) [reply test/sanbase/auth/user_test.exs]: You're right, just changed it

File: lib/sanbase_web/graphql/resolvers/user/user_resolver.ex (hunk 8 above)
Comment (spiderjako) [reply lib/sanbase_web/graphql/resolvers/user/user_resolver.ex]: Alright

File: lib/sanbase_web/graphql/schema/queries/user_queries.ex [hunk 10]
```diff
@@ -226,5 +226,13 @@ defmodule SanbaseWeb.Graphql.Schema.UserQueries do
       middleware(JWTAuth)
       resolve(&UserSettingsResolver.update_user_settings/3)
     end
+
+    @desc ~s"Initiate change of the user's avatar."
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema/queries/user_queries.ex]: why `initiate`? It's a full one-step change

Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/schema/queries/user_queries.ex]: Also no need for the sigil `~s`.
Off-topic - you're using sigils most often when you have quotes `"` inside the string and don't want to escape them manually. So to use sigils with `"` separators is not very useful. Throughout our codebase we use mainly `~s/`

File: test/sanbase_web/graphql/user/user_api_test.exs [hunk 11]
```diff
@@ -552,4 +552,25 @@ defmodule SanbaseWeb.Graphql.UserApiTest do
...
+  test "change avatar of current user", %{conn: conn} do
+    mock(Sanbase.MandrillApi, :send, {:ok, %{}})
+
+    new_avatar =
+      "http://stage-sanbase-images.s3.amazonaws.com/uploads/_empowr-coinHY5QG72SCGKYWMN4AEJQ2BRDLXNWXECT.png"
+
+    query = """
+    mutation {
+      changeAvatar(avatar_url: "#{new_avatar}") {
+        avatar_url
```
Comment (IvanIvanoff) [inline test/sanbase_web/graphql/user/user_api_test.exs]: `avatarUrl` for consistency

File: test/sanbase_web/graphql/user/user_api_test.exs [hunk 12]
```diff
@@ -552,4 +552,25 @@ defmodule SanbaseWeb.Graphql.UserApiTest do
     assert json_response(result, 200)["data"]["logout"]["success"]
     assert result.private.plug_session_info == :drop
   end
+
+  test "change avatar of current user", %{conn: conn} do
```
Comment (IvanIvanoff) [inline test/sanbase_web/graphql/user/user_api_test.exs]: Add a mutation test for invalid url, too

File: lib/sanbase_web/graphql/schema/queries/user_queries.ex [hunk 13]
```diff
@@ -226,5 +226,13 @@ defmodule SanbaseWeb.Graphql.Schema.UserQueries do
       middleware(JWTAuth)
       resolve(&UserSettingsResolver.update_user_settings/3)
     end
+
+    @desc "Change of the user's avatar."
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/schema/queries/user_queries.ex]:
```suggestion
    @desc "Change the user's avatar."
```

File: test/sanbase_web/graphql/user/user_api_test.exs [hunk 14]
```diff
@@ -552,4 +552,52 @@ defmodule SanbaseWeb.Graphql.UserApiTest do
...
+    mock(Sanbase.MandrillApi, :send, {:ok, %{}})
```
Comment (tspenov) [inline test/sanbase_web/graphql/user/user_api_test.exs]: this mock is not needed IMO

File: lib/sanbase/auth/user.ex [hunk 15]
```diff
@@ -410,6 +420,20 @@ defmodule Sanbase.Auth.User do
...
+    user
+    |> changeset(%{avatar_url: avatar_url})
+    |> Repo.update()
+    |> case do
+      {:ok, user} ->
+        {:ok, user}
+
+      {:error, changeset} ->
+        {:error,
+         Enum.at(Sanbase.Utils.ErrorHandling.changeset_errors_to_str(changeset).avatar_url, 0)}
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: Check how it's done in other places - do not change the changeset here. Please check how this function is used throughout the codebase

File: test/sanbase_web/graphql/user/user_api_test.exs [hunk 16]
```diff
@@ -552,4 +552,55 @@ defmodule SanbaseWeb.Graphql.UserApiTest do
...
+      "data" => %{"changeAvatar" => nil},
+      "errors" => [
+        %{
+          "message" => message,
+          "details" => details
+        }
+      ]
+    } = json_response(result, 200)
+
+    assert [message, details] ==
```
Comment (IvanIvanoff) [inline test/sanbase_web/graphql/user/user_api_test.exs]: please assert them separately

File: test/sanbase/auth/user_test.exs [hunk 17]
```diff
@@ -400,4 +400,30 @@ defmodule Sanbase.Auth.UserTest do
...
+  end
+
+  test "user with invalid avatar url returns proper error" do
+    avatar_url = "something invalid"
+
+    {:error, [msg, details]} =
+      insert(:user)
+      |> Sanbase.Auth.User.update_avatar_url(avatar_url)
+
+    assert [msg, details] ==
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: assert them sepparately

File: lib/sanbase/auth/user.ex (hunk 15 above)
Comment (spiderjako) [reply lib/sanbase/auth/user.ex]: Should be okay now

File: test/sanbase/auth/user_test.exs [hunk 18]
```diff
@@ -415,13 +415,11 @@ defmodule Sanbase.Auth.UserTest do
...
 
-    {:error, [message: msg, details: details]} =
+    {:error, changeset} =
       insert(:user)
       |> Sanbase.Auth.User.update_avatar_url(avatar_url)
 
-    assert msg == "Cannot change the avatar"
-
-    assert details ==
+    assert errors_on(changeset)[:avatar_url] |> Enum.at(0) ==
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: drop Enum.at and wrap the result in `[]`

Comment (spiderjako) [reply test/sanbase/auth/user_test.exs]: Alright

**Pull request merged**